core/src/main/scala-2/chisel3/FormalContract.scala

```diff
@@ -0,0 +1,35 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3
+
+import chisel3.experimental.SourceInfo
+import scala.language.experimental.macros
+import scala.reflect.macros.whitebox
+
+object FormalContract extends ObjectFormalContractImpl {
+
+  /** Create a `contract` block with one or more arguments and results. */
+  def apply(head: Data, tail: Data*): (Any => Unit) => Any = macro apply_impl
+
+  // This uses a macro internally to allow for contracts with a single argument
+  // and contracts with a tuple of arguments to be created. The macro packages
+  // the sequence of arguments up into a tuple (`Tuple1` for single arguments,
+  // and the corresponding `TupleN` for multiple arguments). Without this macro,
+  // the contracts would always operate on tuples (`Product`), forcing the user
+  // to use
+  //
+  //   val Tuple1(a) = FormalContract(Tuple1(b)) { case Tuple1(x) => ... }
+  //
+  // for contracts with only a single argument.
+  /** @group Macro
+    * @groupprio Macro 1001
+    */
```

Comment:
```suggestion
  /** @group Macro
    * @groupprio Macro 1001
    */
```

perhaps https://github.com/chipsalliance/chisel/blob/57ab35eb940224f1fd93651d87b83842c416db70/macros/src/main/scala/chisel3/SourceInfoDoc.scala#L22 ?

Reply:
I've managed to hide the macros properly thanks to @jackkoenig's new scala-2/3 pattern.